src/dim_helpers/DenseConvDims.jl

```diff
@@ -1,87 +1,103 @@
 export DenseConvDims
 
-"""
-    DenseConvDims
+struct DenseConvDims{N, K, S, P, D}
+    input_size::NTuple{N, Int}
```

Comment:
```suggestion
struct DenseConvDims{N, K, S, P, D, F} <: ConvDims{N, S, P, D, F}
    input_size::NTuple{N, Int}
```
With this and my other suggestion on `flipkernel`, I think it should be possible to keep `DenseConvDims` a subtype of `ConvDims` while still maintaining type stability?

Reply:
Alternatively:
```suggestion
struct DenseConvDims{N, K, S, P, D} <: ConvDims{N, S, P, D, Bool}
```
I would favour the first version over this, however.

Reply:
If we are ok with these parameters having different meaning.
For example in `DenseConvDims` parameter `S` is just the size of the stride tuple, while in `ConvDims` it is the stride tuple itself.

Reply:
Ah, I missed that. How about `struct DenseConvDims{N, K, S, P, D, F} <: ConvDims{Dims{N}, Dims{S}, Dims{P}, Dims{D}, F}` (`Dims{N} === NTuple{N, Int}`)?

Reply:
If I understand you correctly, you still want to encode only type information?
Because right now (before this PR) parameters in  `ConvDims` and `DenseConvDims` are not types, but actual values.
So for the `S` parameter they would be not `NTuple{2, Int}` but `(2, 2)`, for example.

```julia
julia> x = zeros(Float32, 28, 28, 1, 5);
julia> w = rand(Float32, 3, 3, 1, 1);
julia> cdims = DenseConvDims(x, w)
DenseConvDims: (28, 28, 1) * (3, 3) -> (26, 26, 1), stride: (1, 1), pad: (0, 0, 0, 0), dil: (1, 1), flip: false, groups: 1
julia> typeof(cdims)
DenseConvDims{2, (3, 3), 1, 1, 1, (1, 1), (0, 0, 0, 0), (1, 1), false}
```

Therefore structurally, new `DenseConvDims` cannot have the same meaning for its type's parameters, because it does not encode actual values (like `(2, 2)` for the stride), but only type information (like size of the tuple).

Reply:
This is what I get for not properly reading through the codebase first :facepalm:, thanks for clarifying. Please disregard all my comments and LMK when this is ready to go!

Reply:
Not a problem! After getting done with this PR I plan to work on fixing type-stability issues with other conv dims (that would include pooling and depthwise dims). So these `Union`s will not be there for long.